{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f2e7b048",
   "metadata": {},
   "source": [
    "#       InSPIRe Vaginal Resistome Jupyter Notebook (1): processing & normalization\n",
    "\n",
    "**Author : Nassim Boutouchent** \n",
    "\n",
    "Hello! In this first notebook we process data prior to downstream analyses. \n",
    "It mainly focuses on (i) normalization of Antibiotic Resistance genes (ARGs) and Mobile Genetic Elements (MGE) frequency tables using the RPKM method ; and (ii) processing data for (correlation) sparCC analysis and vizualisation.\n",
    "\n",
    "**Note**: *For microbial species we used the previously DESeq2 normilized frequency table for downstream analysis (see Baud et al., 2023,https://doi.org/10.1038/s41598-023-36126-z).*\n",
    "\n",
    "**External Dataset cross-validation** \n",
    "**dataset information** :\n",
    "Feehily, C., Crosby, D., Walsh, C.J. et al.\n",
    "Shotgun sequencing of the vaginal microbiome reveals both a species and functional potential signature of preterm birth.\n",
    "#npj Biofilms Microbiomes 6, 50 (2020). https://doi.org/10.1038/s41522-020-00162-8 \\\n",
    "Data has been uploaded from the European Nucleotide Archive (ENA) under the study accession number PRJEB34536.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fbc0ba7",
   "metadata": {},
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7bbc58c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "310bd2e3",
   "metadata": {},
   "source": [
    "## Data load "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fe8d6aab",
   "metadata": {},
   "outputs": [],
   "source": [
    "## InSPIRE ARGs MGEs counts table\n",
    "arg_count_raw = pd.read_csv(\"../data/Raw/inspire_ARG_count_RF_Raw.csv\", index_col= \"SampleID\")\n",
    "mge_count_raw = pd.read_csv(\"../data/Raw/inspire_MGE_count_Raw.csv\", index_col= \"SampleID\")\n",
    "\n",
    "## External Dataset ARGs counts table\n",
    "external_arg_count_raw = pd.read_csv(\"../data/Raw/ExtSet_ARGs_count_RF_Raw.csv\", index_col=\"SampleID\")\n",
    "\n",
    "##ARGs from CARD db\n",
    "card_arg_count_raw = pd.read_csv(\"../data/Raw/inspire_ARG_count_CARD_Raw.csv\")\n",
    "\n",
    "## load gene lengths for RPK Norm \n",
    "ResFinderdb_gene_lengths = pd.read_csv(\"../data/supp_materials/ARG_ResFinder_v4.3.3_lenght.csv\")\n",
    "MGE_Custommgedb_gene_lengths = pd.read_csv(\"../data/supp_materials/MGEs_FINAL_99perc_trim_length.csv\")\n",
    "CARDdb_gene_lengths = pd.read_csv(\"../data/supp_materials/ARG_CARD_v3.3.0_length.csv\")\n",
    "\n",
    "# InSPIRe final species table (containing the final set of samples)\n",
    "taxonomy_rabun = pd.read_csv(\"../data/Norm/inspire_taxonomy_rabun_sp.csv\", index_col=0)\n",
    "\n",
    "## External dataset final species table (containing all samples)\n",
    "external_taxonomy_rabun = pd.read_csv(\"../data/Norm/ExternalDataset_taxonomy_rabun_sp.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4a366754",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>erm(B)_18_X66468</th>\n",
       "      <th>erm(B)_10_U86375</th>\n",
       "      <th>erm(X)_3_U21300</th>\n",
       "      <th>tet(W)_1_DQ060146</th>\n",
       "      <th>erm(A)_2_AF002716</th>\n",
       "      <th>mef(A)_2_U83667</th>\n",
       "      <th>tet(M)_5_U58985</th>\n",
       "      <th>lsa(C)_1_HM990671</th>\n",
       "      <th>msr(D)_2_AF274302</th>\n",
       "      <th>tet(M)_13_AM990992</th>\n",
       "      <th>cfxA3_1_AF472622</th>\n",
       "      <th>blaTEM-116_1_AY425988</th>\n",
       "      <th>tet(M)_4_X75073</th>\n",
       "      <th>tet(O/W)_5_AM889122</th>\n",
       "      <th>tet(Q)_4_Z21523</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NMH78</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>73</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH64</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>768</td>\n",
       "      <td>390</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH50</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>524</td>\n",
       "      <td>247</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH63</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>457</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH11</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>549</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          erm(B)_18_X66468  erm(B)_10_U86375  erm(X)_3_U21300  \\\n",
       "SampleID                                                        \n",
       "NMH78                    0                 0                0   \n",
       "NMH64                    0                 0                0   \n",
       "NMH50                    0                 0                0   \n",
       "NMH63                    0                 0                0   \n",
       "NMH11                    0                 0                0   \n",
       "\n",
       "          tet(W)_1_DQ060146  erm(A)_2_AF002716  mef(A)_2_U83667  \\\n",
       "SampleID                                                          \n",
       "NMH78                     0                  0                0   \n",
       "NMH64                     0                  0                0   \n",
       "NMH50                     0                  0                0   \n",
       "NMH63                     0                  0                0   \n",
       "NMH11                     0                  0                0   \n",
       "\n",
       "          tet(M)_5_U58985  lsa(C)_1_HM990671  msr(D)_2_AF274302  \\\n",
       "SampleID                                                          \n",
       "NMH78                 105                 73                  0   \n",
       "NMH64                 768                390                  0   \n",
       "NMH50                 524                247                  0   \n",
       "NMH63                   0                457                  0   \n",
       "NMH11                   0                549                  0   \n",
       "\n",
       "          tet(M)_13_AM990992  cfxA3_1_AF472622  blaTEM-116_1_AY425988  \\\n",
       "SampleID                                                                \n",
       "NMH78                      0                 0                      0   \n",
       "NMH64                      0                 0                      0   \n",
       "NMH50                      0                 0                      0   \n",
       "NMH63                      0                 0                      0   \n",
       "NMH11                      0                62                      0   \n",
       "\n",
       "          tet(M)_4_X75073  tet(O/W)_5_AM889122  tet(Q)_4_Z21523  \n",
       "SampleID                                                         \n",
       "NMH78                   0                    0                0  \n",
       "NMH64                   0                    0                0  \n",
       "NMH50                   0                    0                0  \n",
       "NMH63                   0                    0                0  \n",
       "NMH11                   0                    0               81  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#initial formating \n",
    "external_arg_count_raw.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12b5017f",
   "metadata": {},
   "source": [
    "## (1) RPKM Norm "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4ce68e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RPKM Normalization function\n",
    "def compute_rpkm(count_table, geneID_lenght, gene_col=\"Gene\", length_col=\"Length\"):\n",
    "    \"\"\"\n",
    "    RPKM = (10⁹ x C) / (N x L)\n",
    "    C = number of reads mapped to a gene\n",
    "    N = total mapped reads in the sample \n",
    "    L = gene lenght in bp \n",
    "    \"\"\"\n",
    "    gene_length_dict = geneID_lenght.set_index(gene_col)[length_col].to_dict()\n",
    "    count_table = count_table.astype(int)\n",
    "    \n",
    "    rpk_df = count_table.div([gene_length_dict[col] / 1000 for col in count_table.columns], axis=1)\n",
    "    total_reads_per_sample = count_table.sum(axis=1)\n",
    "    scaling_factor = total_reads_per_sample / 1e6\n",
    "\n",
    "    rpkm_df = rpk_df.div(scaling_factor, axis=0)\n",
    "\n",
    "    return rpkm_df\n",
    "\n",
    "# Format table at gene level\n",
    "# initial formating : gene_variant_accessionnumber  -> gene\n",
    "def format_columns(rpkm_count_table, n=0): \n",
    "    rpkm_count_table_gene = rpkm_count_table.copy()\n",
    "    rpkm_count_table_gene.columns = [col.split('_')[n] for col in rpkm_count_table_gene.columns]\n",
    "    rpkm_count_table_gene = rpkm_count_table_gene.groupby(axis=1, level=0).sum()\n",
    "\n",
    "    return rpkm_count_table_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cf187ce",
   "metadata": {},
   "source": [
    "### Antibiotic resistance genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "067ced95",
   "metadata": {},
   "source": [
    "#### from RF DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "813aaac9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      Gene  Length\n",
      "0      aac(6')-Ib_2_M23634     606\n",
      "1  aac(6')-Ib11_1_AY136758     570\n",
      "          erm(B)_18_X66468  erm(B)_10_U86375  erm(X)_3_U21300  \\\n",
      "SampleID                                                        \n",
      "NMH78                    0                 0                0   \n",
      "NMH64                    0                 0                0   \n",
      "\n",
      "          tet(W)_1_DQ060146  erm(A)_2_AF002716  mef(A)_2_U83667  \\\n",
      "SampleID                                                          \n",
      "NMH78                     0                  0                0   \n",
      "NMH64                     0                  0                0   \n",
      "\n",
      "          tet(M)_5_U58985  lsa(C)_1_HM990671  msr(D)_2_AF274302  \\\n",
      "SampleID                                                          \n",
      "NMH78                 105                 73                  0   \n",
      "NMH64                 768                390                  0   \n",
      "\n",
      "          tet(M)_13_AM990992  cfxA3_1_AF472622  blaTEM-116_1_AY425988  \\\n",
      "SampleID                                                                \n",
      "NMH78                      0                 0                      0   \n",
      "NMH64                      0                 0                      0   \n",
      "\n",
      "          tet(M)_4_X75073  tet(O/W)_5_AM889122  tet(Q)_4_Z21523  \n",
      "SampleID                                                         \n",
      "NMH78                   0                    0                0  \n",
      "NMH64                   0                    0                0  \n"
     ]
    }
   ],
   "source": [
    "# initial formating RF\n",
    "print(ResFinderdb_gene_lengths.head(2))\n",
    "print(external_arg_count_raw.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "03c8ae99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>erm(B)_18_X66468</th>\n",
       "      <th>erm(B)_10_U86375</th>\n",
       "      <th>erm(X)_3_U21300</th>\n",
       "      <th>tet(W)_1_DQ060146</th>\n",
       "      <th>erm(A)_2_AF002716</th>\n",
       "      <th>mef(A)_2_U83667</th>\n",
       "      <th>tet(M)_5_U58985</th>\n",
       "      <th>lsa(C)_1_HM990671</th>\n",
       "      <th>msr(D)_2_AF274302</th>\n",
       "      <th>tet(M)_13_AM990992</th>\n",
       "      <th>cfxA3_1_AF472622</th>\n",
       "      <th>blaTEM-116_1_AY425988</th>\n",
       "      <th>tet(M)_4_X75073</th>\n",
       "      <th>tet(O/W)_5_AM889122</th>\n",
       "      <th>tet(Q)_4_Z21523</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NMH78</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>307233.146067</td>\n",
       "      <td>277290.303956</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH64</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>345423.143351</td>\n",
       "      <td>227713.025535</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH50</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>353977.518374</td>\n",
       "      <td>216607.954511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH63</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>676132.521974</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>536411.495035</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92748.836152</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60774.674518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          erm(B)_18_X66468  erm(B)_10_U86375  erm(X)_3_U21300  \\\n",
       "SampleID                                                        \n",
       "NMH78                  0.0               0.0              0.0   \n",
       "NMH64                  0.0               0.0              0.0   \n",
       "NMH50                  0.0               0.0              0.0   \n",
       "NMH63                  0.0               0.0              0.0   \n",
       "NMH11                  0.0               0.0              0.0   \n",
       "\n",
       "          tet(W)_1_DQ060146  erm(A)_2_AF002716  mef(A)_2_U83667  \\\n",
       "SampleID                                                          \n",
       "NMH78                   0.0                0.0              0.0   \n",
       "NMH64                   0.0                0.0              0.0   \n",
       "NMH50                   0.0                0.0              0.0   \n",
       "NMH63                   0.0                0.0              0.0   \n",
       "NMH11                   0.0                0.0              0.0   \n",
       "\n",
       "          tet(M)_5_U58985  lsa(C)_1_HM990671  msr(D)_2_AF274302  \\\n",
       "SampleID                                                          \n",
       "NMH78       307233.146067      277290.303956                0.0   \n",
       "NMH64       345423.143351      227713.025535                0.0   \n",
       "NMH50       353977.518374      216607.954511                0.0   \n",
       "NMH63            0.000000      676132.521974                0.0   \n",
       "NMH11            0.000000      536411.495035                0.0   \n",
       "\n",
       "          tet(M)_13_AM990992  cfxA3_1_AF472622  blaTEM-116_1_AY425988  \\\n",
       "SampleID                                                                \n",
       "NMH78                    0.0          0.000000                    0.0   \n",
       "NMH64                    0.0          0.000000                    0.0   \n",
       "NMH50                    0.0          0.000000                    0.0   \n",
       "NMH63                    0.0          0.000000                    0.0   \n",
       "NMH11                    0.0      92748.836152                    0.0   \n",
       "\n",
       "          tet(M)_4_X75073  tet(O/W)_5_AM889122  tet(Q)_4_Z21523  \n",
       "SampleID                                                         \n",
       "NMH78                 0.0                  0.0         0.000000  \n",
       "NMH64                 0.0                  0.0         0.000000  \n",
       "NMH50                 0.0                  0.0         0.000000  \n",
       "NMH63                 0.0                  0.0         0.000000  \n",
       "NMH11                 0.0                  0.0     60774.674518  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# InSPIRe ARGs RPKM Norm \n",
    "arg_count_rpkm = compute_rpkm(arg_count_raw, ResFinderdb_gene_lengths, gene_col=\"Gene\", length_col=\"Length\")\n",
    "\n",
    "# External Dataset ARGs RPKM Norm\n",
    "external_arg_count_rpkm = compute_rpkm(external_arg_count_raw, ResFinderdb_gene_lengths, gene_col=\"Gene\", length_col=\"Length\")\n",
    "external_arg_count_rpkm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d2de3a26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ARR-3', 'NarA', 'NarB', 'OqxA', 'OqxB', 'aac(3)-IIa', 'aac(3)-IId', 'aac(3)-IV', 'aac(3)-Id', 'aac(3)-XI']\n",
      "['blaTEM-116', 'cfxA3', 'erm(A)', 'erm(B)', 'erm(X)', 'lsa(C)', 'mef(A)', 'msr(D)', 'tet(M)', 'tet(O/W)']\n"
     ]
    }
   ],
   "source": [
    "# format columns to remove accession numbers and group by gene names \n",
    "arg_count_rpkm_gene = format_columns(arg_count_rpkm, n =0)\n",
    "external_arg_count_rpkm_gene = format_columns(external_arg_count_rpkm, n =0)\n",
    "\n",
    "print(arg_count_rpkm_gene.columns.tolist()[:10])\n",
    "print(external_arg_count_rpkm_gene.columns.tolist()[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "add9f633",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ant(6)-Ia_5_AB247327</th>\n",
       "      <th>msr(A)_2_AB013298</th>\n",
       "      <th>aph(3'')-Ib_5_AF321551</th>\n",
       "      <th>blaZ_133_AIDT01000014</th>\n",
       "      <th>blaZ_117_ACHE01000004</th>\n",
       "      <th>blaZ_132_AHVO01000010</th>\n",
       "      <th>fusB_1_AY373761</th>\n",
       "      <th>cat(pC194)_1_NC_002013</th>\n",
       "      <th>sul3_2_AJ459418</th>\n",
       "      <th>sul1_36_X15024</th>\n",
       "      <th>...</th>\n",
       "      <th>vga(A)LC_2_DQ823382</th>\n",
       "      <th>dfrA1_9_AJ238350</th>\n",
       "      <th>blaLEN7_1_AJ635425</th>\n",
       "      <th>blaSHV-110_1_HQ877615</th>\n",
       "      <th>dfrA1_5_EU089668</th>\n",
       "      <th>tet(32)_2_EF626943</th>\n",
       "      <th>aadA1b_1_M95287</th>\n",
       "      <th>cfxA3_1_AF472622</th>\n",
       "      <th>blaTEM-182_1_HQ317449</th>\n",
       "      <th>msr(C)_2_AF313494</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IMVM_0001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0002</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0004</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0006</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0008</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 349 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           ant(6)-Ia_5_AB247327  msr(A)_2_AB013298  aph(3'')-Ib_5_AF321551  \\\n",
       "SampleID                                                                     \n",
       "IMVM_0001                   0.0                0.0                     0.0   \n",
       "IMVM_0002                   0.0                0.0                     0.0   \n",
       "IMVM_0004                   0.0                0.0                     0.0   \n",
       "IMVM_0006                   0.0                0.0                     0.0   \n",
       "IMVM_0008                   0.0                0.0                     0.0   \n",
       "\n",
       "           blaZ_133_AIDT01000014  blaZ_117_ACHE01000004  \\\n",
       "SampleID                                                  \n",
       "IMVM_0001                    0.0                    0.0   \n",
       "IMVM_0002                    0.0                    0.0   \n",
       "IMVM_0004                    0.0                    0.0   \n",
       "IMVM_0006                    0.0                    0.0   \n",
       "IMVM_0008                    0.0                    0.0   \n",
       "\n",
       "           blaZ_132_AHVO01000010  fusB_1_AY373761  cat(pC194)_1_NC_002013  \\\n",
       "SampleID                                                                    \n",
       "IMVM_0001                    0.0              0.0                     0.0   \n",
       "IMVM_0002                    0.0              0.0                     0.0   \n",
       "IMVM_0004                    0.0              0.0                     0.0   \n",
       "IMVM_0006                    0.0              0.0                     0.0   \n",
       "IMVM_0008                    0.0              0.0                     0.0   \n",
       "\n",
       "           sul3_2_AJ459418  sul1_36_X15024  ...  vga(A)LC_2_DQ823382  \\\n",
       "SampleID                                    ...                        \n",
       "IMVM_0001              0.0             0.0  ...                  0.0   \n",
       "IMVM_0002              0.0             0.0  ...                  0.0   \n",
       "IMVM_0004              0.0             0.0  ...                  0.0   \n",
       "IMVM_0006              0.0             0.0  ...                  0.0   \n",
       "IMVM_0008              0.0             0.0  ...                  0.0   \n",
       "\n",
       "           dfrA1_9_AJ238350  blaLEN7_1_AJ635425  blaSHV-110_1_HQ877615  \\\n",
       "SampleID                                                                 \n",
       "IMVM_0001               0.0                 0.0                    0.0   \n",
       "IMVM_0002               0.0                 0.0                    0.0   \n",
       "IMVM_0004               0.0                 0.0                    0.0   \n",
       "IMVM_0006               0.0                 0.0                    0.0   \n",
       "IMVM_0008               0.0                 0.0                    0.0   \n",
       "\n",
       "           dfrA1_5_EU089668  tet(32)_2_EF626943  aadA1b_1_M95287  \\\n",
       "SampleID                                                           \n",
       "IMVM_0001               0.0                 0.0              0.0   \n",
       "IMVM_0002               0.0                 0.0              0.0   \n",
       "IMVM_0004               0.0                 0.0              0.0   \n",
       "IMVM_0006               0.0                 0.0              0.0   \n",
       "IMVM_0008               0.0                 0.0              0.0   \n",
       "\n",
       "           cfxA3_1_AF472622  blaTEM-182_1_HQ317449  msr(C)_2_AF313494  \n",
       "SampleID                                                               \n",
       "IMVM_0001               0.0                    0.0                0.0  \n",
       "IMVM_0002               0.0                    0.0                0.0  \n",
       "IMVM_0004               0.0                    0.0                0.0  \n",
       "IMVM_0006               0.0                    0.0                0.0  \n",
       "IMVM_0008               0.0                    0.0                0.0  \n",
       "\n",
       "[5 rows x 349 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print final formating \n",
    "arg_count_rpkm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "339bb318",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1957, 231)\n",
      "(89, 12)\n"
     ]
    }
   ],
   "source": [
    "# Keep negative samples, as they have important biological meaning\n",
    "inspire_samples = taxonomy_rabun.index.tolist()\n",
    "arg_count_rpkm_gene = arg_count_rpkm_gene.reindex(inspire_samples, fill_value=0)\n",
    "\n",
    "# Also keep negative samples for the External dataset\n",
    "ExternalSet_samples = external_taxonomy_rabun.index.tolist()\n",
    "external_arg_count_rpkm_gene = external_arg_count_rpkm_gene.reindex(ExternalSet_samples, fill_value=0)\n",
    "\n",
    "# check \n",
    "print(arg_count_rpkm_gene.shape)\n",
    "print(external_arg_count_rpkm_gene.shape)\n",
    "\n",
    "#7 samples, from patients who withdrew their consent to the study, has been excluded for downstream analysis. \n",
    "#['IMVM_0217', 'IMVM_0712', 'IMVM_1200', 'IMVM_1304', 'IMVM_1607', 'IMVM_1794', 'IMVM_1841']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40340e11",
   "metadata": {},
   "source": [
    "#### from CARD  db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ac7ef4e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Gene Name  Accession                                    AMR Gene Family  \\\n",
      "0      mdtA    3000792  resistance-nodulation-cell division (RND) anti...   \n",
      "1      evgA    3000832  major facilitator superfamily (MFS) antibiotic...   \n",
      "\n",
      "                                          Drug Class  IMVM_0001  IMVM_0002  \\\n",
      "0                           aminocoumarin antibiotic      920.0        0.0   \n",
      "1  macrolide antibiotic; fluoroquinolone antibiot...      486.0        0.0   \n",
      "\n",
      "   IMVM_0003  IMVM_0004  IMVM_0005  IMVM_0006  ...  IMVM_1975  IMVM_1976  \\\n",
      "0          0        0.0        0.0        0.0  ...        0.0        0.0   \n",
      "1          0        0.0        0.0        0.0  ...        0.0        0.0   \n",
      "\n",
      "   IMVM_1977  IMVM_1978  IMVM_1979  IMVM_1980  IMVM_1981  IMVM_1982  \\\n",
      "0        0.0        0.0        0.0     1036.0        0.0        0.0   \n",
      "1        0.0        5.0        0.0      904.0        0.0        0.0   \n",
      "\n",
      "   IMVM_1983  IMVM_1984  \n",
      "0        0.0       22.0  \n",
      "1        0.0        6.0  \n",
      "\n",
      "[2 rows x 1963 columns]\n",
      "  Gene Name   ARO ID     NCBI ID  Gene Length\n",
      "0    CblA-1  3002999  GQ343019.1          891\n",
      "1    SHV-52  3001109  HQ845196.1          861\n"
     ]
    }
   ],
   "source": [
    "# initial formating CARD\n",
    "print(card_arg_count_raw.head(2))\n",
    "print(CARDdb_gene_lengths.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a7302667",
   "metadata": {},
   "outputs": [],
   "source": [
    "#format\n",
    "card_arg_count_raw.rename(columns={\"Accession\": \"ARO ID\"}, inplace=True)\n",
    "col_to_drop = ['AMR Gene Family', 'Drug Class','Gene Name','Gene Name']\n",
    "card_arg_count_raw_ARO = card_arg_count_raw.copy()\n",
    "card_arg_count_raw_ARO = card_arg_count_raw_ARO.set_index(\"ARO ID\")\n",
    "card_arg_count_raw_ARO = card_arg_count_raw_ARO.drop(columns=col_to_drop).T\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e9b715c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# InSPIRe ARGs RPKM Norm \n",
    "card_arg_count_rpkm_ARO = compute_rpkm(card_arg_count_raw_ARO, CARDdb_gene_lengths, gene_col=\"ARO ID\", length_col=\"Gene Length\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2d263537",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['mdtA', 'evgA', 'mdtE', 'lsaA', 'emrB', 'tet(X)', 'H-NS', 'CRP', 'emrY', 'mel']\n"
     ]
    }
   ],
   "source": [
    "# Format to gene count table\n",
    "ARO_to_gene = CARDdb_gene_lengths.set_index('ARO ID')['Gene Name'].to_dict()\n",
    "card_arg_count_rpkm_gene = card_arg_count_rpkm_ARO.rename(columns=ARO_to_gene)\n",
    "print(card_arg_count_rpkm_gene.columns.tolist()[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "181a831c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1957, 492)\n"
     ]
    }
   ],
   "source": [
    "inspire_samples = taxonomy_rabun.index.tolist()\n",
    "card_arg_count_rpkm_gene = card_arg_count_rpkm_gene.reindex(inspire_samples, fill_value=0).fillna(0)\n",
    "print(card_arg_count_rpkm_gene.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13da3b9d",
   "metadata": {},
   "source": [
    "### MGEs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5bc65dac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      Header  Gene Length\n",
      "0        414_tnpA_AL513383.1         3497\n",
      "1   2570_tnpA_CCSD01000073.1         3090\n",
      "2       2004_tnpA_HF677572.1         3084\n",
      "3  733_tnpA1_CBTR010000003.1         3081\n",
      "4   1017_tnpA_FKLR01000027.1         3066\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1002_tnpA_JQ364968.1</th>\n",
       "      <th>1004_tnpA_KT779035.1</th>\n",
       "      <th>1017_tnpA_FKLR01000027.1</th>\n",
       "      <th>101_rep20_10_repA(SAP057A)_NC_013334.1</th>\n",
       "      <th>1020_IS91_LFAQ01000019.1</th>\n",
       "      <th>102_rep20_11_repA(VRSAp)_AP003367</th>\n",
       "      <th>1037_tnpA_AY849557.2</th>\n",
       "      <th>103_rep21_1_rep(pWBG754)_GQ900396.1</th>\n",
       "      <th>1041_IS91_AIGD01000018.1</th>\n",
       "      <th>1044_tnpA_AB723628.1</th>\n",
       "      <th>...</th>\n",
       "      <th>954_tnpA_KP987218.1</th>\n",
       "      <th>955_tnpA_JQ318854.1</th>\n",
       "      <th>956_tnpA7_HG916853.1</th>\n",
       "      <th>957_tnpA_KX029332.1</th>\n",
       "      <th>968_tnpA_Z17279.1</th>\n",
       "      <th>978_tnpA_AF141323.1</th>\n",
       "      <th>982_tnpA_JX296013.1</th>\n",
       "      <th>98_rep20_7_repA(SAP074A)_GQ900426.1</th>\n",
       "      <th>997_tnpA_AF342826.1</th>\n",
       "      <th>999_tnpA_rve_CEGG01000033.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IMVM_0001</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0002</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0003</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0004</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0005</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 764 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           1002_tnpA_JQ364968.1  1004_tnpA_KT779035.1  \\\n",
       "SampleID                                                \n",
       "IMVM_0001                     0                     0   \n",
       "IMVM_0002                     0                     0   \n",
       "IMVM_0003                     0                     0   \n",
       "IMVM_0004                     0                     0   \n",
       "IMVM_0005                     0                     0   \n",
       "\n",
       "           1017_tnpA_FKLR01000027.1  101_rep20_10_repA(SAP057A)_NC_013334.1  \\\n",
       "SampleID                                                                      \n",
       "IMVM_0001                         0                                       0   \n",
       "IMVM_0002                         0                                       0   \n",
       "IMVM_0003                         0                                       0   \n",
       "IMVM_0004                         0                                       0   \n",
       "IMVM_0005                         0                                       0   \n",
       "\n",
       "           1020_IS91_LFAQ01000019.1  102_rep20_11_repA(VRSAp)_AP003367  \\\n",
       "SampleID                                                                 \n",
       "IMVM_0001                         0                                  0   \n",
       "IMVM_0002                         0                                  0   \n",
       "IMVM_0003                         0                                  0   \n",
       "IMVM_0004                         0                                  0   \n",
       "IMVM_0005                         0                                  0   \n",
       "\n",
       "           1037_tnpA_AY849557.2  103_rep21_1_rep(pWBG754)_GQ900396.1  \\\n",
       "SampleID                                                               \n",
       "IMVM_0001                     0                                    0   \n",
       "IMVM_0002                     0                                    0   \n",
       "IMVM_0003                     0                                    0   \n",
       "IMVM_0004                     0                                    0   \n",
       "IMVM_0005                     0                                    0   \n",
       "\n",
       "           1041_IS91_AIGD01000018.1  1044_tnpA_AB723628.1  ...  \\\n",
       "SampleID                                                   ...   \n",
       "IMVM_0001                         0                     0  ...   \n",
       "IMVM_0002                         0                     0  ...   \n",
       "IMVM_0003                         0                     0  ...   \n",
       "IMVM_0004                         0                     0  ...   \n",
       "IMVM_0005                         0                     0  ...   \n",
       "\n",
       "           954_tnpA_KP987218.1  955_tnpA_JQ318854.1  956_tnpA7_HG916853.1  \\\n",
       "SampleID                                                                    \n",
       "IMVM_0001                    0                    0                     0   \n",
       "IMVM_0002                    0                    0                     0   \n",
       "IMVM_0003                    0                    0                     0   \n",
       "IMVM_0004                    0                    0                     0   \n",
       "IMVM_0005                    0                    0                     0   \n",
       "\n",
       "           957_tnpA_KX029332.1  968_tnpA_Z17279.1  978_tnpA_AF141323.1  \\\n",
       "SampleID                                                                 \n",
       "IMVM_0001                    0                  0                    0   \n",
       "IMVM_0002                    0                  0                    0   \n",
       "IMVM_0003                    0                  0                    0   \n",
       "IMVM_0004                    0                  0                    0   \n",
       "IMVM_0005                    0                  0                    0   \n",
       "\n",
       "           982_tnpA_JX296013.1  98_rep20_7_repA(SAP074A)_GQ900426.1  \\\n",
       "SampleID                                                              \n",
       "IMVM_0001                    0                                    0   \n",
       "IMVM_0002                    0                                    0   \n",
       "IMVM_0003                    0                                    0   \n",
       "IMVM_0004                    0                                    0   \n",
       "IMVM_0005                    0                                    0   \n",
       "\n",
       "           997_tnpA_AF342826.1  999_tnpA_rve_CEGG01000033.1  \n",
       "SampleID                                                     \n",
       "IMVM_0001                    0                            0  \n",
       "IMVM_0002                    0                            0  \n",
       "IMVM_0003                    0                            0  \n",
       "IMVM_0004                    0                            0  \n",
       "IMVM_0005                    0                            0  \n",
       "\n",
       "[5 rows x 764 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#initial formating \n",
    "print(MGE_Custommgedb_gene_lengths.head())\n",
    "mge_count_raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ea9aaf95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>erm(B)_18_X66468</th>\n",
       "      <th>erm(B)_10_U86375</th>\n",
       "      <th>erm(X)_3_U21300</th>\n",
       "      <th>tet(W)_1_DQ060146</th>\n",
       "      <th>erm(A)_2_AF002716</th>\n",
       "      <th>mef(A)_2_U83667</th>\n",
       "      <th>tet(M)_5_U58985</th>\n",
       "      <th>lsa(C)_1_HM990671</th>\n",
       "      <th>msr(D)_2_AF274302</th>\n",
       "      <th>tet(M)_13_AM990992</th>\n",
       "      <th>cfxA3_1_AF472622</th>\n",
       "      <th>blaTEM-116_1_AY425988</th>\n",
       "      <th>tet(M)_4_X75073</th>\n",
       "      <th>tet(O/W)_5_AM889122</th>\n",
       "      <th>tet(Q)_4_Z21523</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NMH78</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>307233.146067</td>\n",
       "      <td>277290.303956</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH64</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>345423.143351</td>\n",
       "      <td>227713.025535</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH50</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>353977.518374</td>\n",
       "      <td>216607.954511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH63</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>676132.521974</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMH11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>536411.495035</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92748.836152</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60774.674518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          erm(B)_18_X66468  erm(B)_10_U86375  erm(X)_3_U21300  \\\n",
       "SampleID                                                        \n",
       "NMH78                  0.0               0.0              0.0   \n",
       "NMH64                  0.0               0.0              0.0   \n",
       "NMH50                  0.0               0.0              0.0   \n",
       "NMH63                  0.0               0.0              0.0   \n",
       "NMH11                  0.0               0.0              0.0   \n",
       "\n",
       "          tet(W)_1_DQ060146  erm(A)_2_AF002716  mef(A)_2_U83667  \\\n",
       "SampleID                                                          \n",
       "NMH78                   0.0                0.0              0.0   \n",
       "NMH64                   0.0                0.0              0.0   \n",
       "NMH50                   0.0                0.0              0.0   \n",
       "NMH63                   0.0                0.0              0.0   \n",
       "NMH11                   0.0                0.0              0.0   \n",
       "\n",
       "          tet(M)_5_U58985  lsa(C)_1_HM990671  msr(D)_2_AF274302  \\\n",
       "SampleID                                                          \n",
       "NMH78       307233.146067      277290.303956                0.0   \n",
       "NMH64       345423.143351      227713.025535                0.0   \n",
       "NMH50       353977.518374      216607.954511                0.0   \n",
       "NMH63            0.000000      676132.521974                0.0   \n",
       "NMH11            0.000000      536411.495035                0.0   \n",
       "\n",
       "          tet(M)_13_AM990992  cfxA3_1_AF472622  blaTEM-116_1_AY425988  \\\n",
       "SampleID                                                                \n",
       "NMH78                    0.0          0.000000                    0.0   \n",
       "NMH64                    0.0          0.000000                    0.0   \n",
       "NMH50                    0.0          0.000000                    0.0   \n",
       "NMH63                    0.0          0.000000                    0.0   \n",
       "NMH11                    0.0      92748.836152                    0.0   \n",
       "\n",
       "          tet(M)_4_X75073  tet(O/W)_5_AM889122  tet(Q)_4_Z21523  \n",
       "SampleID                                                         \n",
       "NMH78                 0.0                  0.0         0.000000  \n",
       "NMH64                 0.0                  0.0         0.000000  \n",
       "NMH50                 0.0                  0.0         0.000000  \n",
       "NMH63                 0.0                  0.0         0.000000  \n",
       "NMH11                 0.0                  0.0     60774.674518  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RPKM Morn\n",
    "mge_count_rpkm = compute_rpkm(mge_count_raw, MGE_Custommgedb_gene_lengths, gene_col=\"Header\", length_col=\"Gene Length\")\n",
    "external_arg_count_rpkm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cca8b81c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Col(BS512)', 'Col(IRGK)', 'Col(MG828)', 'Col(MP18)', 'Col156', 'Col3M', 'Col8282', 'ColE10', 'ColRNAI', 'ColpVC']\n",
      "(1957, 157)\n"
     ]
    }
   ],
   "source": [
    "#Format columns \n",
    "mge_count_rpkm_gene = format_columns(mge_count_rpkm, n=1)\n",
    "print(mge_count_rpkm_gene.columns.tolist()[:10])\n",
    "\n",
    "# Keep also negative samples for MGEs \n",
    "mge_count_rpkm_gene = mge_count_rpkm_gene.reindex(inspire_samples, fill_value=0)\n",
    "print(mge_count_rpkm_gene.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4b6c74a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save RPKM tables for downstream analysis\n",
    "#mge_count_rpkm_gene.to_csv(\"Norm/MGE_count_RF_RPKM_Gene.csv\")\n",
    "#arg_count_rpkm_gene.to_csv(\"Norm/ARG_count_RF_RPKM_Gene.csv\")\n",
    "#external_arg_count_rpkm_gene.to_csv(\"Norm/ExtSet_ARGs_count_RF_RPKM_Gene.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffd089e0",
   "metadata": {},
   "source": [
    "## (2) SparCC analysis  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cddcdd48",
   "metadata": {},
   "source": [
    "### Process data prior to analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "86cbb17a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "top 20 species: ['Lactobacillus_crispatus', 'Lactobacillus_iners', 'Gardnerella_vaginalis', 'Lactobacillus_jensenii', 'Enterococcus_faecalis', 'Lactobacillus_gasseri', 'Escherichia_coli', 'Bifidobacterium_breve', 'Gardnerella_leopoldii', 'Fannyhessea_vaginae', 'Klebsiella_pneumoniae', 'Staphylococcus_aureus', 'Prevotella_bivia', 'Gardnerella_swidsinskii', 'Gardnerella_piotii', 'Lactobacillus_paragasseri', 'Streptomyces_gancidicus', 'Streptococcus_agalactiae', 'Candida_albicans', 'Prevotella_timonensis']\n",
      "top 50 ARGs: ['lsa(C)', 'tet(M)', 'erm(B)', 'lsa(A)', \"aph(3')-III\", 'cfxA3', 'mef(A)', 'msr(D)', 'blaTEM-1B', 'erm(A)', 'tet(Q)', 'erm(X)', 'lnu(C)', 'erm(F)', 'ant(6)-Ia', 'blaTEM-116', \"aph(3'')-Ib\", 'mre(A)', 'aph(6)-Id', 'fosA', 'sul2', 'mph(C)', 'fosB', 'blaMAL-1', 'tet(W)', 'tet(O)', 'tet(L)', 'tet(A)', 'fosA7', 'erm(C)', 'blaZ', 'cfxA', 'dfrA1', 'dfrA17', 'sul1', 'tet(B)', 'tet(O/32/O)', 'blaTEM-1A', 'OqxA', 'lnu(A)', 'catA2', 'OqxB', 'cfxA4', 'bleO', 'dfrG', 'cat', 'msr(A)', 'tet(O/W)', 'blaTEM-1C', 'qnrS1']\n"
     ]
    }
   ],
   "source": [
    "# Keep the 20 most abundant species \n",
    "top20_sp = taxonomy_rabun.sum().sort_values(ascending=False).head(20).index.tolist()\n",
    "\n",
    "# Keep also the 50 most abundant ARGs\n",
    "top50_args = arg_count_rpkm_gene.sum().sort_values(ascending=False).head(50).index.tolist()\n",
    "\n",
    "#diplay lists \n",
    "print(f'top 20 species: {top20_sp}')\n",
    "print(f'top 50 ARGs: {top50_args}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "956e3a16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "top 50 MGEs: ['Tn916', 'tnpA', 'Xis-Tn916', 'ColRNAI', 'IS91', 'Int-Tn916', 'Col156', 'Col(MG828)', 'tnpA1', 'rep6', 'Col(BS512)', 'rep11', 'rep9', 'rep10', 'int2', 'tnpAB', 'IS621', 'tnpA4', 'IncFII(S)', 'tnpAIS1', 'rep21', 'repUS2', 'ColpVC', 'delta-tnpA', 'istB', 'int3', 'IncFIB(AP001918)', 'Col8282', 'rep7', 'repUS11', 'rep20', 'rep18', 'tnpA3', 'rep8', 'IncFIA', 'IncI1', 'rep5', 'tnpA10', 'intI1', 'IncFII(pCRY)', 'tnpA5', 'qacEdelta', 'rep2', 'IncQ1', 'istA2', 'IncFII(29)', 'rep10b', 'tniA', 'IncX1', 'IncFII']\n"
     ]
    }
   ],
   "source": [
    "# ORFs belonging to the same Tn916 were grouped prior to MGEs filtering\n",
    "mge_count_rpkm_sparcc_cleaned = mge_count_rpkm_gene.copy()\n",
    "\n",
    "cols = [Tn for Tn in mge_count_rpkm_sparcc_cleaned.columns if Tn.startswith(\"Tn916-\")]\n",
    "mge_count_rpkm_sparcc_cleaned[\"Tn916\"] = mge_count_rpkm_sparcc_cleaned[cols].sum(axis=1)\n",
    "mge_count_rpkm_sparcc_cleaned.drop(columns=cols, inplace=True)\n",
    "\n",
    "# Keep also the 50 most abundant MGEs\n",
    "top50_mges = mge_count_rpkm_sparcc_cleaned.sum().sort_values(ascending=False).head(50).index.tolist()\n",
    "print (f'top 50 MGEs: {top50_mges}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "042de4db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "merged matrice ARGs+sp shape: (1957, 70)\n",
      "merged matrice ARGs+sp+MGEs shape: (1957, 120)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lsa(C)</th>\n",
       "      <th>tet(M)</th>\n",
       "      <th>erm(B)</th>\n",
       "      <th>lsa(A)</th>\n",
       "      <th>aph(3')-III</th>\n",
       "      <th>cfxA3</th>\n",
       "      <th>mef(A)</th>\n",
       "      <th>msr(D)</th>\n",
       "      <th>blaTEM-1B</th>\n",
       "      <th>erm(A)</th>\n",
       "      <th>...</th>\n",
       "      <th>tnpA5</th>\n",
       "      <th>qacEdelta</th>\n",
       "      <th>rep2</th>\n",
       "      <th>IncQ1</th>\n",
       "      <th>istA2</th>\n",
       "      <th>IncFII(29)</th>\n",
       "      <th>rep10b</th>\n",
       "      <th>tniA</th>\n",
       "      <th>IncX1</th>\n",
       "      <th>IncFII</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IMVM_0001</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>98329.122646</td>\n",
       "      <td>136932.926518</td>\n",
       "      <td>33139.314918</td>\n",
       "      <td>115559.17641</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6788.390043</td>\n",
       "      <td>3451.383189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>188257.264848</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11367.548245</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0002</th>\n",
       "      <td>676132.521974</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0003</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0004</th>\n",
       "      <td>425963.488844</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>301837.270341</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>191256.830601</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMVM_0005</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  lsa(C)        tet(M)         erm(B)        lsa(A)  \\\n",
       "SampleID                                                              \n",
       "IMVM_0001       0.000000  98329.122646  136932.926518  33139.314918   \n",
       "IMVM_0002  676132.521974      0.000000       0.000000      0.000000   \n",
       "IMVM_0003       0.000000      0.000000       0.000000      0.000000   \n",
       "IMVM_0004  425963.488844      0.000000  301837.270341      0.000000   \n",
       "IMVM_0005       0.000000      0.000000       0.000000      0.000000   \n",
       "\n",
       "            aph(3')-III  cfxA3       mef(A)       msr(D)  blaTEM-1B  \\\n",
       "SampleID                                                              \n",
       "IMVM_0001  115559.17641    0.0  6788.390043  3451.383189        0.0   \n",
       "IMVM_0002       0.00000    0.0     0.000000     0.000000        0.0   \n",
       "IMVM_0003       0.00000    0.0     0.000000     0.000000        0.0   \n",
       "IMVM_0004       0.00000    0.0     0.000000     0.000000        0.0   \n",
       "IMVM_0005       0.00000    0.0     0.000000     0.000000        0.0   \n",
       "\n",
       "                  erm(A)  ...  tnpA5  qacEdelta          rep2  IncQ1  istA2  \\\n",
       "SampleID                  ...                                                 \n",
       "IMVM_0001  188257.264848  ...    0.0        0.0  11367.548245    0.0    0.0   \n",
       "IMVM_0002       0.000000  ...    0.0        0.0      0.000000    0.0    0.0   \n",
       "IMVM_0003       0.000000  ...    0.0        0.0      0.000000    0.0    0.0   \n",
       "IMVM_0004  191256.830601  ...    0.0        0.0      0.000000    0.0    0.0   \n",
       "IMVM_0005       0.000000  ...    0.0        0.0      0.000000    0.0    0.0   \n",
       "\n",
       "           IncFII(29)  rep10b  tniA  IncX1  IncFII  \n",
       "SampleID                                            \n",
       "IMVM_0001         0.0     0.0   0.0    0.0     0.0  \n",
       "IMVM_0002         0.0     0.0   0.0    0.0     0.0  \n",
       "IMVM_0003         0.0     0.0   0.0    0.0     0.0  \n",
       "IMVM_0004         0.0     0.0   0.0    0.0     0.0  \n",
       "IMVM_0005         0.0     0.0   0.0    0.0     0.0  \n",
       "\n",
       "[5 rows x 120 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Formmat table to fit the input requirements of the fastspar tool\n",
    "# (i) spcies & ARGs \n",
    "matrice_top50ARGs = arg_count_rpkm_gene[top50_args]\n",
    "matrice_top20sp = taxonomy_rabun[top20_sp]\n",
    "merge_matrice_ARGsp = matrice_top50ARGs.merge(matrice_top20sp, left_index= True, right_index= True)\n",
    "print(f'merged matrice ARGs+sp shape: {merge_matrice_ARGsp.shape}')\n",
    "\n",
    "# (ii) species & ARGs & MGEs\n",
    "matrice_top50MGEs = mge_count_rpkm_sparcc_cleaned[top50_mges]\n",
    "merge_matrice_ARGspMGE = merge_matrice_ARGsp.merge(matrice_top50MGEs, left_index= True, right_index= True)\n",
    "print(f'merged matrice ARGs+sp+MGEs shape: {merge_matrice_ARGspMGE.shape}')\n",
    "merge_matrice_ARGspMGE.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be7afc0a",
   "metadata": {},
   "source": [
    "#### Compute a compositionnaly aware correlation matrix using Fastspar (https://github.com/scwatts/fastspar.git)\n",
    "\n",
    "**(1) Correlation inference**\n",
    "```{bash} \n",
    "fastspar --otu_table merge_matrice_ARGsp.tsv --correlation median_correlation.tsv --covariance median_covariance.tsv\n",
    "```\n",
    "**(2) bootstrap**\n",
    "```{bash} \n",
    "mkdir bootstrap_counts\n",
    "fastspar_bootstrap --otu_table merge_matrice_ARGsp.tsv --number 1000 --prefix bootstrap_counts/merge_matrice_ARGsp\n",
    "```\n",
    "\n",
    "**(3) Infer correlation** \n",
    "```{bash} \n",
    "mkdir bootstrap_correlation\n",
    "parallel fastspar --otu_table {} --correlation bootstrap_correlation/cor_{/} --covariance bootstrap_correlation/cov_{/} -i 5 ::: bootstrap_counts/*\n",
    "```\n",
    "**final p-values table**\n",
    "```{bash} \n",
    "fastspar_pvalues --otu_table merge_matrice_ARGsp.tsv --correlation median_correlation.tsv --prefix bootstrap_correlation/merge_matrice_ARGsp --permutations 1000 --outfile pvalues.tsv\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ba54723",
   "metadata": {},
   "source": [
    "### Post-processing of FastSpar outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dae658d",
   "metadata": {},
   "source": [
    "#### load output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2fd48b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load fastspar output (corr and p-value matrix) (supplementary data tables)\n",
    "\n",
    "#ARG+sp\n",
    "SPARGs_median_corr = pd.read_csv(\"../data/Output/CorAbun_20Sp50ARGs.csv\", sep=';', index_col='#OTU ID') \n",
    "SPARGs_pvalues_corr = pd.read_csv(\"../data/Output/CorAbun_20Sp50ARGs_pval.csv\", sep=';', index_col='#OTU ID')\n",
    "\n",
    "#ARG+sp+MGE\n",
    "SPARGsMGEs_median_corr = pd.read_csv(\"../data/Output/CorAbun_20Sp50ARG50MGE.csv\", sep=';', index_col='#OTU ID') \n",
    "SPARGsMGEs_pvalues_corr = pd.read_csv(\"../data/Output/CorAbun_20Sp50ARG50MGE_pval.csv\", sep=';', index_col='#OTU ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e5dff0d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert a correlation matrix to a df long format for network vizualisation in Gephi\n",
    "def convert_to_network(corr_matrix, value_name=\"Weight\"):\n",
    "    network_df = (corr_matrix.stack().reset_index().rename(columns={'#OTU ID': 'Source', 'level_1': 'Target', 0: value_name}))\n",
    "    network_df = network_df[network_df[\"Source\"] != network_df[\"Target\"]] # Remove self-loops\n",
    "    # Remove duplicate edges (A–B == B–A)\n",
    "    network_df[\"pair\"] = network_df.apply(lambda row: tuple(sorted([row[\"Source\"], row[\"Target\"]])), axis=1)\n",
    "    network_df = network_df.drop_duplicates(subset=\"pair\").drop(columns=\"pair\")\n",
    "\n",
    "    return network_df\n",
    "\n",
    "# Filter network dataframe based on correlation and p-value thresholds\n",
    "def filter_network(network, corr=0.1, pvalue=0.01):\n",
    "    filter_criteria = (network['Weight'] > corr) & (network['pvalue'] < pvalue)\n",
    "    network_filtered = network[filter_criteria]\n",
    "    \n",
    "    return network_filtered\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "aba1cf89",
   "metadata": {},
   "outputs": [],
   "source": [
    "SPARGs_network_weight = convert_to_network(SPARGs_median_corr, \"Weight\")\n",
    "SPARGs_network_pvalue = convert_to_network(SPARGs_pvalues_corr, \"pvalue\")\n",
    "SPARGsMGEs_network_weight = convert_to_network(SPARGsMGEs_median_corr, \"Weight\")\n",
    "SPARGsMGEs_network_pvalue = convert_to_network(SPARGsMGEs_pvalues_corr, \"pvalue\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1544cb2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>Target</th>\n",
       "      <th>Weight</th>\n",
       "      <th>pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Lactobacillus_crispatus</td>\n",
       "      <td>Lactobacillus_gasseri</td>\n",
       "      <td>0.3786</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Lactobacillus_crispatus</td>\n",
       "      <td>Lactobacillus_jensenii</td>\n",
       "      <td>0.4049</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>Escherichia_coli</td>\n",
       "      <td>Enterococcus_faecalis</td>\n",
       "      <td>0.1900</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Escherichia_coli</td>\n",
       "      <td>Citrobacter_koseri</td>\n",
       "      <td>0.4352</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Escherichia_coli</td>\n",
       "      <td>Klebsiella_pneumoniae</td>\n",
       "      <td>0.5021</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Source                  Target  Weight  pvalue\n",
       "6   Lactobacillus_crispatus   Lactobacillus_gasseri  0.3786   0.001\n",
       "7   Lactobacillus_crispatus  Lactobacillus_jensenii  0.4049   0.001\n",
       "69         Escherichia_coli   Enterococcus_faecalis  0.1900   0.001\n",
       "72         Escherichia_coli      Citrobacter_koseri  0.4352   0.001\n",
       "73         Escherichia_coli   Klebsiella_pneumoniae  0.5021   0.001"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# spARGs Network (corr > 1 and p-value < 0.01)\n",
    "network_ARGsp = SPARGs_network_weight.merge(SPARGs_network_pvalue[[\"Source\", \"Target\", \"pvalue\"]], on=[\"Source\", \"Target\"], how=\"inner\")\n",
    "network_ARGsp_filtered = filter_network(network_ARGsp, corr=0.1, pvalue=0.01)   \n",
    "\n",
    "# spARGsMGEs Network (corr > 1 and p-value < 0.01)\n",
    "network_ARGspMGE = SPARGsMGEs_network_weight.merge(SPARGsMGEs_network_pvalue[[\"Source\", \"Target\", \"pvalue\"]], on=[\"Source\", \"Target\"], how=\"inner\")\n",
    "network_ARGspMGE_filtered = filter_network(network_ARGspMGE, corr=0.1, pvalue=0.01) \n",
    "\n",
    "#Print example\n",
    "network_ARGsp_filtered.head()\n",
    "\n",
    "# These dfs are used as inputs for Gephi (see Methods for network visualization)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mon_environnement",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
